core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -0,0 +1,283 @@
+
+    errorHandler.accept(errorResponse);
+
+    // Throw an exception in case the provided error handler does not throw.
+    throw new RESTException("Unhandled error: %s", errorResponse);
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param requestBody  - Content to place in the request body
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
+   * @return The response entity, parsed and converted to its type T
+   */
+  private <T> T execute(
+      Method method, String path, Object requestBody, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    if (path.startsWith("/")) {
+      throw new RESTException(
+          "Received a malformed path for a REST request: %s. Paths should not start with /", path);
+    }
+
+    String fullUri = String.format("%s/%s", uri, path);
+    HttpUriRequestBase request = new HttpUriRequestBase(method.name(), URI.create(fullUri));
+    addRequestHeaders(request);
+
+    if (requestBody != null) {
+      try {
+        StringEntity stringEntity = new StringEntity(mapper.writeValueAsString(requestBody));
+        request.setEntity(stringEntity);
+      } catch (JsonProcessingException e) {
+        throw new RESTException(e, "Failed to write request body: %s", requestBody);
+      }
+    }
+
+    try (CloseableHttpResponse response = httpClient.execute(request)) {
+
+      // Skip parsing the response stream for any successful request not expecting a response body
+      if (responseType == null && isSuccessful(response)) {
+        return null;
+      }
+
+      String responseBody = extractResponseBodyAsString(response);
+
+      if (!isSuccessful(response)) {
+        // The provided error handler is expected to throw, but a RESTException is thrown if not.
+        throwFailure(response, responseBody, errorHandler);
+      }
+
+      if (responseBody == null) {
+        throw new RESTException(
+            "Invalid (null) response body for request (expected %s): method=%s, path=%s, status=%d",
+            responseType.getSimpleName(), method.name(), path, response.getCode());
+      }
+
+      try {
+        return mapper.readValue(responseBody, responseType);
+      } catch (JsonProcessingException e) {
+        throw new RESTException(
+            "Received a success response code of %d, but failed to parse response body into %s",
+            response.getCode(), responseType.getSimpleName());
+      }
+    } catch (IOException e) {
+      throw new RESTException(e, "Error occurred while processing %s request", method);
```

Comment:
I just merged this, but realized that I think we need to change the handling here. If `execute` throws an `IOException`, it is caught here, so it needs to be handled by the (an?) error handler so that the error handler can decide how to propagate the exception.

For example, if the request is updating a table and times out, then `IOException` / `InterruptedByTimeoutException` might be thrown. In that case, the handler needs to throw `CommitStateUnknownException` because it doesn't know whether the request completed successfully.

Reply:
Oh that's a good catch @rdblue.

I understand the issue with catching `IOException`, as the error handler could potentially throw it, but I don't quite understand your suggested fix.

I've been sick this weekend and so it might be really clear and i'm just not grokking your comment.